### ISSUE TITLE
Server doesn't start after upgrade to v1.21.7+k3s1 (ac705709): bootstrap data already found and encrypted with different token

### ISSUE THREAD

**Issue by zdzichu** [k3s-io/k3s#4640]
**Environmental Info:**
K3s Version: v1.21.7+k3s1 (ac705709)

Node(s) CPU architecture, OS, and Version: 

5.15.4-201.fc35.x86_64 #1 SMP Tue Nov 23 18:54:50 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux
Fedora Linux 35

Cluster Configuration:

I have 2 servers using external PostgreSQL database and couple of agents.
There's system-upgrade-controller configured to follow `channel: https://update.k3s.io/v1-release/channels/stable`

**Describe the bug:**

Today auto upgrade from **v1.21.5+k3s2** to **v1.21.7+k3s1** happened on on server. It stopped responding. Service logs shows following failure:
```
k3s[1644879]: time="2021-12-04T11:17:49.379252883+01:00" level=info msg="Starting k3s v1.21.7+k3s1 (ac705709)"
k3s[1644879]: time="2021-12-04T11:17:49.420438206+01:00" level=info msg="Configuring postgres database connection pooling: maxIdleConns=2, maxOpenConns=0, connMaxLifetime=0s"
k3s[1644879]: time="2021-12-04T11:17:49.420639983+01:00" level=info msg="Configuring database table schema and indexes, this may take a moment..."
k3s[1644879]: time="2021-12-04T11:17:49.423536941+01:00" level=info msg="Database tables and indexes are up to date"
k3s[1644879]: time="2021-12-04T11:17:49.448156983+01:00" level=info msg="Kine listening on unix://kine.sock"
k3s[1644879]: time="2021-12-04T11:17:49.480807423+01:00" level=fatal msg="starting kubernetes: preparing server: bootstrap data already found and encrypted with different token"
systemd[1]: k3s.service: Main process exited, code=exited, status=1/FAILURE
```

I have verified `/var/lib/rancher/k3s/server/token` is the same on both servers.
After **downgrading** k3s binary back to v1.21.5+k3s2, server is able to start again.

**Steps To Reproduce:**

- Installed K3s: 
Cluster installed 342 days ago using:
export INSTALL_K3S_COMMIT=fadc5a8057c244df11757cd47cc50cc4a4cf5887
export K3S_DATASTORE_ENDPOINT='postgres://[…]'
./k3s-install server \
 	--tls-san api.chi.pipebreaker.pl \
	--disable traefik \
	--flannel-backend=wireguard 

Then it was auto-upgraded using `system-upgrade-controller`, stable channel, up to v1.21.5+k3s2.

**Expected behavior:**

Server should start after upgrade to v1.21.7+k3s1

**Actual behavior:**

Startup failed with `level=fatal msg="starting kubernetes: preparing server: bootstrap data already found and encrypted with different token"`

Does handling the bootstrap data encryption changed between 1.21.5+k3s2 and 1.21.7+k3s1?

**Additional context / logs:**

There are corresponding logs from Postgresql:
```
2021-12-04 11:05:52.461 CET [2243563] ERROR:  relation "key_value" does not exist at character 22
2021-12-04 11:05:52.461 CET [2243563] STATEMENT:  SELECT COUNT(*) FROM key_value
2021-12-04 11:05:52.511 CET [2243567] LOG:  could not receive data from client: Connection reset by peer
2021-12-04 11:05:52.511 CET [2243566] LOG:  could not send data to client: Connection reset by peer
2021-12-04 11:05:52.511 CET [2243566] STATEMENT:  
                                SELECT (
                        SELECT MAX(rkv.id) AS id
                        FROM kine AS rkv), (
                        SELECT MAX(crkv.prev_revision) AS prev_revision
                        FROM kine AS crkv
                        WHERE crkv.name = 'compact_rev_key'), kv.id AS theid, kv.name, kv.created, kv.deleted, kv.create_revision, kv.prev_revision, kv.lease, kv.value, kv.old_value
                                FROM kine AS kv
                                WHERE
                                        kv.name LIKE $1 AND
                                        kv.id > $2
                                ORDER BY kv.id ASC LIMIT 500
2021-12-04 11:05:52.511 CET [2243565] LOG:  could not send data to client: Connection reset by peer

2021-12-04 11:05:52.511 CET [2243565] STATEMENT:  
                        SELECT (
                        SELECT MAX(rkv.id) AS id
                        FROM kine AS rkv), (
                        SELECT MAX(crkv.prev_revision) AS prev_revision
                        FROM kine AS crkv
                        WHERE crkv.name = 'compact_rev_key'), kv.id AS theid, kv.name, kv.created, kv.deleted, kv.create_revision, kv.prev_revision, kv.lease, kv.value, kv.old_value
                        FROM kine AS kv
                        JOIN (
                                SELECT MAX(mkv.id) AS id
                                FROM kine AS mkv
                                WHERE
                                        mkv.name LIKE $1
                                        
                                GROUP BY mkv.name) maxkv
                    ON maxkv.id = kv.id
                        WHERE
                                  (kv.deleted = 0 OR $2)
                        ORDER BY kv.id ASC
                         LIMIT 1000
2021-12-04 11:05:52.511 CET [2243566] FATAL:  connection to client lost
```

**Backporting**

- [ ] Needs backporting to older releases

**Comment by brandond** [comment]
It looks like you've not specified a token on either of the nodes? Use of a token for HA with external SQL became required as of https://github.com/k3s-io/k3s/releases/tag/v1.21.3+k3s1

>⚠️ Important Upgrade Note ⚠️
If you are using K3s in a HA configuration with an external SQL datastore, and your server (control-plane) nodes were not started with the --token CLI flag, you will no longer be able to add additional K3s servers to the cluster without specifying the token. Ensure that you retain a copy of this token, as is required when restoring from backup. Previously, K3s did not enforce the use of a token when using external SQL datastores.
>
>You may retrieve the token value from any server already joined to the cluster:
>
>`cat /var/lib/rancher/k3s/server/token`

If you've already upgraded, and are affected by this issue, you should copy the token from the server that was upgraded *first* to the other servers.

**Comment by zdzichu** [comment]
I saw this note, but I'm not adding any new master, so it did not seem relevant to me.

`/var/lib/rancher/k3s/server/token` is the same on both of my servers.
When I added it to `k3s.service` (which seems terribly insecure, but nevertheless):

```
ExecStart=/usr/local/bin/k3s \
    server \
        '--tls-san' \
        'api.chi.pipebreaker.pl' \
        '--disable' \
        'traefik' \
        '--flannel-backend=wireguard' \
        '--token' \
        'K[…]::server:1d[…]'
```
I get the same failure (bootstrap data encrypted with different token) with **v1.21.5+k3s2**, which was working previously!
After I remove `--token …` from unit file, 1.21.5 start again.

**Comment by zdzichu** [comment]
I managed to start v1.21.5+k3s2 with token provided. I've created `/etc/rancher/k3s/config.yaml` with:
```yaml
---
token: ""
```
(empty value). This allows my current configuration to start.
But after changing `k3s` binary to v1.21.7+k3s1, the error is the same.
With .7, I've tried changing `token:` to:
- `""`
- `K[…]::server:1d[…]` (from /var/lib/rancher/k3s/server/token)
- `server:1d[…]`
- `1d[…]`
Neither works with v1.21.7+k3s1

**Comment by briandowns** [comment]
Is it possible to post the output from this query? `select * from kine where name like "%bootstrap%";`

**Comment by zdzichu** [comment]
Sure,
```
k3s> select * from kine where name like '/bootstrap%';                            
+-----------+-------------------------+-----------+-----------+-------------------+-----------------+---------+----------------------------------------------------->
| id        | name                    | created   | deleted   | create_revision   | prev_revision   | lease   | value                                               >
|-----------+-------------------------+-----------+-----------+-------------------+-----------------+---------+----------------------------------------------------->
| 130375029 | /bootstrap/4489e5a9bd37 | 1         | 0         | 0                 | 0               | 0       | \x393262663131333764656465363165363a504e475179327a57>
+-----------+-------------------------+-----------+-----------+-------------------+-----------------+---------+----------------------------------------------------
```

I know this number should be first 12 bytes of sha256 of server password from `/var/lib/rancher/k3s/server/cred/passwd`, but it doesn't fit. I didn't find any string which would hash to this value.

**Comment by zdzichu** [comment]
Ah sorry, you wanted one with % at the beginning, too
```
k3s> select * from kine where name like '%bootstrap%';                             
+-----------+-----------------------------------------------------------------------+-----------+-----------+-------------------+-----------------+---------+-------->
| id        | name                                                                  | created   | deleted   | create_revision   | prev_revision   | lease   | value  >
|-----------+-----------------------------------------------------------------------+-----------+-----------+-------------------+-----------------+---------+-------->
| 99        | /registry/clusterroles/system:node-bootstrapper                       | 1         | 0         | 0                 | 0               | 0       | \x6b387>
| 183       | /registry/roles/kube-public/system:controller:bootstrap-signer        | 1         | 0         | 0                 | 0               | 0       | \x6b387>
| 185       | /registry/roles/kube-system/system:controller:bootstrap-signer        | 1         | 0         | 0                 | 0               | 0       | \x6b387>
| 193       | /registry/rolebindings/kube-system/system:controller:bootstrap-signer | 1         | 0         | 0                 | 0               | 0       | \x6b387>
| 196       | /registry/rolebindings/kube-public/system:controller:bootstrap-signer | 1         | 0         | 0                 | 0               | 0       | \x6b387>
| 130375029 | /bootstrap/4489e5a9bd37                                               | 1         | 0         | 0                 | 0               | 0       | \x39326>
+-----------+-----------------------------------------------------------------------+-----------+-----------+-------------------+-----------------+---------+-------->
SELECT 6
```

**Comment by galal-hussein** [comment]
the number `4489e5a9bd37` is the hash of the token you specified in the /var/lib/rancher/k3s/server/token, which should correctly match the token you specified originally, I will take a deeper look into this and try to reproduce the problem if possible with postgress or mysql

**Comment by zdzichu** [comment]
@galal-hussein when trying to reproduce, keep in mind:
- I've initially installed version 1.19 (first commit adding support for cgroupfs2)
- I did **not** specify a token during installation
- cluster has been upgraded using `stable` channel

When I remove  /var/lib/rancher/k3s/server/token and restart k3s, the file is recreated with content of `K[…]::server:1d[…]`. Yet when I put this value as `token:`, k3s won't start.
I wonder: when the token is empty (`""`), what value k3s hashes to get /bootstrap/*hash*?

**Comment by brandond** [comment]
> I wonder: when the token is empty (""), what value k3s hashes to get /bootstrap/hash?

This is discussed in https://github.com/k3s-io/k3s/security/advisories/GHSA-cxm9-4m6p-24mc

**Comment by zdzichu** [comment]
@brandond 
Started with:
export INSTALL_K3S_COMMIT=fadc5a8057c244df11757cd47cc50cc4a4cf5887

As for
> On startup, updated versions of K3s will automatically re-encrypt the bootstrap data using the cluster token, and delete the copy encrypted with the incorrect key.

It doesn't seem to be true:

>  $ sha256sum  /var/lib/rancher/k3s/server/token
> 95db1cbf24f32543477622103a2f82d7b231bf9bed83f474caa555327f6378db  /var/lib/rancher/k3s/server/token

Which does not contain 4489e5a9bd37.

**Comment by galal-hussein** [comment]
I am trying to reproduce the problem, but upgrading from 1.21.5 to 1.21.7 didnt reveal any issues, I will try the path starting from 1.19 to 1.21.5 and then 1.217

as for the token this is the function that we use for getting the bootstrap token hash
https://github.com/k3s-io/k3s/blob/master/pkg/cluster/encrypt.go#L26-L30

but you are correct it doesnt look like the token that you specified and its not empty token either

**Comment by galal-hussein** [comment]
@zdzichu can you try loading your token and using with the function I mentioned above and see if it will result the same name of the bootstrap key

**Comment by brandond** [comment]
>> $ sha256sum /var/lib/rancher/k3s/server/token
>> 95db1cbf24f32543477622103a2f82d7b231bf9bed83f474caa555327f6378db /var/lib/rancher/k3s/server/token
>
> Which does not contain 4489e5a9bd

That's not what it is. Take a look at https://github.com/k3s-io/k3s/blob/7d3447ceff88f7cc2cedc20d187eba80d686814b/pkg/cluster/encrypt.go#L26-L30

From the shell, the equivalent would be:
```bash
echo -n "token" | sha256sum | cut -c -12
```

**Comment by galal-hussein** [comment]
also the commit fadc5a8057c244df11757cd47cc50cc4a4cf5887 doesnt seem to be available anymore on google drive, I will try to use the nearest release to that

**Comment by galal-hussein** [comment]
okay here are my findings:

at version v1.19.5+k3s1, the bootstrap token is saved with empty string:
```
mysql> select name from kine where name like "%bootstrap%";
+-----------------------------------------------------------------------+
| name                                                                  |
+-----------------------------------------------------------------------+
| /bootstrap/e3b0c44298fc                                               |
```

when upgrading to v1.21.5+k3s2 the bootstrap still is an empty token, which shouldnt happen since it should be automatically migrated this key and created a new one which should be properly encrypted
```
mysql> select name from kine where name like "%bootstrap%";
+-----------------------------------------------------------------------+
| name                                                                  |
+-----------------------------------------------------------------------+
| /bootstrap/e3b0c44298fc                                               |
```

After upgrading to v1.21.7 I can see the following in the logs:
```
Dec 07 21:24:22 ubuntu-focal k3s[15501]: time="2021-12-07T21:24:22.207318791Z" level=warning msg="bootstrap data encrypted with empty string, deleting and resaving with token"
Dec 07 21:24:22 ubuntu-focal k3s[15501]: time="2021-12-07T21:24:22.242292727Z" level=info msg="created bootstrap key /bootstrap/e0e255e54ffc"
```
but the weird part is that I can still see the empty string bootstrap key there **TWICE**:
```
mysql> select name from kine where name like "%bootstrap%";
+-----------------------------------------------------------------------+
| name                                                                  |
+-----------------------------------------------------------------------+
| /bootstrap/e0e255e54ffc                                               |
| /bootstrap/e3b0c44298fc                                               |
| /bootstrap/e3b0c44298fc                                               |
| /registry/clusterroles/system:node-bootstrapper                       |

```
cc @briandowns @brandond

**Comment by brandond** [comment]
That's what you want to see if you're selecting directly from the database. The second one is when it gets marked as deleted in the datastore: 
```sql
select name, deleted from kine where name like "/bootstrap/%";`
```

It'll go away after it gets compacted.

**Comment by galal-hussein** [comment]
oh yeah I forgot about the compaction part of k3s, I just wonder why the empty string token didnt get deleted when upgrade happened to 1.21.5

**Comment by brandond** [comment]
Yeah, the fact that it doesn't get migrated when upgrading to v1.21.5+k3s2 is unexpected.

**Comment by galal-hussein** [comment]
@brandond I think in 1.21.5 k3s looks for the stamp file and doesnt enter the bootstrap section at all, due to some legacy code created by Darren:

https://github.com/k3s-io/k3s/blob/v1.21.5%2Bk3s2/pkg/cluster/bootstrap.go#L88-L92

this part was commented in 1.21.7 and I can see the correct behavior being executed in 1.21.7 and the empty string token is being deleted and migrated to a new bootstrap key

**Comment by galal-hussein** [comment]
@zdzichu I cant really reproduce the problem when upgrading to 1.21.7, however I understand why its working with 1.21.5 since 1.21.5 skipped the bootstrap check at all, the issue here is that you should actually get an empty string token since you originally started with 1.19.5 which makes me think that you originally had the --token set ? if not then we need to compare the hash of the token with bootstrap key that you found in the database to make sure its the right one

**Comment by zdzichu** [comment]
The "passphrase" in my case is the string starting with `1d…` (I will not paste it fully on a public issue).
I can find it at:
- last field of `/var/lib/rancher/k3s/server/token`
- first field of both lines (for node and for server) of `/var/lib/rancher/k3s/server/cred/passwd`
The sha256 of this `1d…` string starts with `0188bdd871e8` which is DIFFERENT than `4489e5a9bd37` that I have in the database.

Keep in mind that 1.21.5+k3s2 with empty token provided IS ABLE to find and decrypt /bootstrap/4489e5a9bd37.

**Comment by galal-hussein** [comment]
@zdzichu the code in 1.21.5 actually skipped the decryption of the bootstrap entirely for the first node as I described in this comment https://github.com/k3s-io/k3s/issues/4640#issuecomment-988293710, is there a possiblity that token has been changed anytime between upgrading from 1.19 to 1.21.5 ?

**Comment by zdzichu** [comment]
I've found a script used to install agents a year ago. The token was the same as it is now.

**Comment by brandond** [comment]
We haven't changed the token hash calculation since `v0.11.0-alpha3`.

You mentioned above that you did not specify the token when initially creating the cluster; this means that each node had a *different* value in `/var/lib/rancher/k3s/server/token` when the cluster was upgraded; I believe you need to take manual action to find the node that was upgraded first (the one whose token matches the bootstrap hash) and set that token on the remaining nodes.

**Comment by zdzichu** [comment]
I was able to fix the issue. To summarize:
- 2 servers were initially installed without specifying token (with an empty one)
- the migration code around 1.21.5+k3s2 was slightly broken
- there was a `/bootstrap/4489e5a9bd37` entry in the database
- `/var/lib/rancher/k3s/server/token` on both master has the same content, but it *wasn't used* to get `4489e5a9bd37` shortcut
- it looks like bootstrap data was encrypted using unknown key.

I did the following steps to recover the situation:
- went through database backups and found the most recent `/bootstrap/e3b0c44298fc`; `e3b…` is the result of using an empty token
- stopped k3s on both servers
- replaced the `/bootstrap/4489…` entry in the database with `/bootstrap/e3b0…`
- created `/etc/rancher/k3s/config.yaml` with `token: "<content of /var/lib/rancher/k3s/server/token>"` on both servers
- downloaded k3s-v1.21.8+k3s1
- crossed my fingers and started k3s on one of the servers

At this point k3s started, migration code noticed empty token was used initially, bootstrap data was re-encoded. I have only `/bootstrap/0188bdd871e8` entry in the database, and `0188…` corresponds to sha256 sum of password field in my token.
Second server started without issue.

Thanks for guidance on this issue. I've been able to peek deeply into k3s bootstrap and learned quite a bit.

**Comment by timvandruenen** [comment]
Hi all, we're in the exact same situation as @zdzichu. Following his steps exactly is resulting in the following error:
`level=warning msg="bootstrap data encrypted with empty string, deleting and resaving with token"
level=fatal msg="starting kubernetes: preparing server: cipher: message authentication failed"`

Anybody an idea about what could cause the fatal error? We're kind of stuck on 1.21.5 at the moment.

**Comment by kopwei** [comment]
I have the same issue and now I got stuck at v1.21.1. I first tried to add a new node (v1.21.5) into the cluster (which uses external ETCD without a token) and we realized the problem, then we removed the new node and tried to add nodes with v1.21.1 but got the error "bootstrap data already found and encrypted with different token"

**Comment by timvandruenen** [comment]
@kopwei sadly we didn't manage to fix this. As we only run Rancher on our K3s cluster, we removed the cluster, installed a new K3s version and restored Rancher.

**Comment by superherointj** [comment]
> bootstrap data already found and encrypted with different token

I solved this problem but my issue was different. I had k3s nodes still running when resetting `etcd`, so `etcd` database would get polluted continuously from k3s instances still running in other hosts. Solution was to stop all k3s instances in all hosts first. Then, delete `etcd` database and only then to re-enable k3s. [I did this way because I can scrap the data.]

TL-DR: Check if K3s nodes are running (which should not be) when deleting `etcd` database.
